cpp/src/reductions/scan/ewm.cu

```diff
@@ -0,0 +1,377 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "scan.cuh"
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/scan.h>
+
+namespace cudf {
+namespace detail {
+
+template <typename T>
+class recurrence_functor {
+  using pair_type = thrust::pair<T, T>;
+
+ public:
+  __device__ pair_type operator()(pair_type ci, pair_type cj)
+  {
+    return {ci.first * cj.first, ci.second * cj.first + cj.second};
+  }
+};
+
+/**
+ * @brief Solve a recurrence relation using a blelloch scan
+ * The second elements of the pairs will contain the result
+ */
+template <typename T>
+void compute_recurrence(rmm::device_uvector<thrust::pair<T, T>>& input,
+                        rmm::cuda_stream_view stream)
+{
+  recurrence_functor<T> op;
+  thrust::inclusive_scan(rmm::exec_policy(stream), input.begin(), input.end(), input.begin(), op);
+}
+
+/**
+* @brief Return an array whose values y_i are the number of null entries
+* in between the last valid entry of the input and the current index.
+* Example: {1, NULL, 3, 4, NULL, NULL, 7}
+        -> {0, 0     1, 0, 0,    1,    2}
+*/
+rmm::device_uvector<cudf::size_type> null_roll_up(column_view const& input,
+                                                  rmm::cuda_stream_view stream)
+{
+  rmm::device_uvector<cudf::size_type> output(
+    input.size(), stream, rmm::mr::get_current_device_resource());
+
+  auto device_view = *column_device_view::create(input);
```

Comment:
Don't discard the unique pointer. You should store it as is, then dereference it wherever you need it.